// @eartharoid/vite-plugin-i18n/dist/I18nPlugin.js
import { I18n } from '@eartharoid/i18n';
import { mtoc } from '@eartharoid/cif';
import { createFilter, normalizePath, } from 'vite';
function parse(messages, namespace) {
    const i18n = new I18n({ defer_extraction: false });
    return i18n.parse(messages, namespace);
}
export default function I18nPlugin(options) {
    return {
        enforce: 'pre',
        name: 'i18n',
        transform(src, id) {
            const filter = createFilter(options.include, options.exclude);
            const [normalised, qs] = normalizePath(id).split('?');
            if (filter(normalised)) {
                const query = new URLSearchParams(qs ?? '');
                const id_regex = options.id_regex || /(?<id>[a-z0-9-_]+)\.[a-z]+/i;
                let { locale, namespace } = id_regex.exec(normalised)?.groups || {};
                namespace = query.get('namespace') || namespace;
                const messages = options.parser ? options.parser(src) : JSON.parse(src);
                const parsed = parse(messages, namespace);
                const cif = mtoc(parsed);
                return {
                    code: JSON.stringify({
                        ...options.compact ? { cif } : { json: parsed },
                        locale_id: locale
                    }),
                    map: { mappings: '' },
                };
            }
        },
    };
}
